htetl/tasks/main.py

```diff
@@ -0,0 +1,150 @@
+import luigi
+import logging
+import pandas as pd
+import networkx as nx
+import string
+import htetl.extract.emails
+import htetl.get_data as get_data
+import htetl.reverse_url as reverse_url
+import htetl.tasks.phones
+import htetl.util as util
+
+
+logger = logging.getLogger('luigi-interface')
+logger.setLevel(logging.DEBUG)
+
+
+class MakeGraph(luigi.Task):
+    """  reversing url task """
+
+    NEW_DATA_EXTRACTOR = luigi.BoolParameter(
+        config_path=dict(section='htetl-flags', name='new_data_extractor'),
+        default=False
+    )
+
+    def output(self):
+        out_path = 'data/graph.csv'
+        return luigi.LocalTarget(out_path)
+
+    def requires(self):
+        if self.NEW_DATA_EXTRACTOR:
+            return {
+                'email': htetl.extract.emails.ParseEmails(),
+                'phone': htetl.tasks.phones.ParsePhones(),
+                # TODO: This needs to be adapted to the new Page data
+                # extraction methods like ParseEmail and ParsePhones
+                'oid': get_data.RawPosterData(),
+            }
+        else:
+            return [get_data.RawPhoneData(),
+                    get_data.RawEmailData(),
+                    get_data.RawPosterData()]
+
+    def run(self):
+        """ data is small enough to use pandas for the processing """
+
+        out = []
+
+        # TODO: When htetl-flags:new_data_extractor is removed, remove the
+        # flatten call and clean up the for loop
+        for in_target in luigi.task.flatten(self.input()):
+            with in_target.open('r') as f:
+                data = pd.read_csv(f)
+
+            for i, (k, v) in enumerate(data.groupby(data.columns[-1])):
+                values = v.values.tolist()
+                page_ids = [x[0] for x in values]
+                offset_page_ids = page_ids[1:]
+                if len(page_ids) == 1:
+                    offset_page_ids = page_ids
+                else:
+                    offset_page_ids.append(page_ids[0])
+                out.append([
+                    (a, b)
+                    for a, b in zip(page_ids, offset_page_ids)
+                ])
+
+        out = [item for sublist in out for item in sublist]
+
+        G = nx.Graph()
+        G.add_edges_from(out)
+
+        sub_graphs = []
+        for i, x in enumerate(nx.connected_component_subgraphs(G)):
+            nodes = nx.nodes(x)
+            sub_graphs.append(list(zip([i] * len(nodes), nodes)))
+
+        sub_graphs = [item for sublist in sub_graphs for item in sublist]
```

Comment:
list comprehension! sublist in sub_graph is the subgraph so item in sublist is the node, item (1,node1), (1,node2),(2,node3)... etc where the nodes are the backpagepostids. looks good again